### ISSUE TITLE
Poco::Data::RecordSet row iteration doesn't work when the statment has limit()

### ISSUE THREAD

**Issue by kostya-lnk-ms** [pocoproject/poco#793]
When a statement has a defined limit, the following code doesn't work:

```
selectStmt << "select * from Test ", Poco::Data::Keywords::limit(1);
while (!selectStmt.done())
{
            selectStmt.execute();
        Poco::Data::RecordSet rs(selectStmt);

        prevRowCount = rowCount;
        std::cout << rs.rowCount();

        for (auto rowIt = rs.begin(); rowIt != rs.end(); ++rowIt)
        {
            const Poco::Data::Row& reader = *rowIt;
            std::cout << "Row: " << reader.valuesToString() << std::endl;
        }
}
```

The code doesn't work, since RecordSet implementation uses underlying statement subTotalRowCount() (which accumulates row counts) as an measure of row count, however, every execute causes underlying extractor storage cleanup, hence storage contains no more than 1 row for every execution.

**Comment by aleks-f** [comment]
Here's a patch; SQLite tests pass but it has to be tested with other back-ends before it is committed. If you can spare some time and do the testing, it will speed up the upstreaming.
```
--- a/Data/src/StatementImpl.cpp                                                                                 
+++ b/Data/src/StatementImpl.cpp                                                                                 
@@ -87,9 +87,15 @@ std::size_t StatementImpl::execute(const bool& reset)                                         
        {                                                                                                        
                compile();                                                                                       
                if (_extrLimit.value() == Limit::LIMIT_UNLIMITED)                                                
+               {                                                                                                
                        lim += executeWithoutLimit();                                                            
+                       assignSubTotal(false);                                                                   
+               }                                                                                                
                else                                                                                             
+               {                                                                                                
                        lim += executeWithLimit();                                                               
+                       assignSubTotal(true);                                                                    
+               }                                                                                                
        } while (canCompile());                                                                                  
                                                                                                                 
        if (_extrLimit.value() == Limit::LIMIT_UNLIMITED)                                                        
@@ -98,8 +104,6 @@ std::size_t StatementImpl::execute(const bool& reset)                                         
        if (lim < _lowerLimit)                                                                                   
                throw LimitException("Did not receive enough data.");                                            
                                                                                                                 
-       assignSubTotal(reset);                                                                                   
-                                                                                                                
        return lim;                                                                                              
 }                                                                                                               
                                                                                                                 
@@ -115,9 +119,9 @@ void StatementImpl::assignSubTotal(bool reset)                                               
                        if (_extractors[counter].size())                                                         
                        {                                                                                        
                                if (reset)                                                                       
-                                       *it += CountVec::value_type(_extractors[counter][0]->numOfRowsHandled());
-                               else                                                                             
                                        *it = CountVec::value_type(_extractors[counter][0]->numOfRowsHandled()); 
+                               else                                                                             
+                                       *it += CountVec::value_type(_extractors[counter][0]->numOfRowsHandled());
                        }                                                                                        
                }                                                                                                
        }                                                                                                        
```

**Comment by kostya-lnk-ms** [comment]
Aleks, thanks for quick response. I just realized, that things become much more complicated when multiple result sets involved as well. e.g. when underlying query consists of more than 1 select.
I've got that fixed myself, however, my fix doesn't account for row filtering - if one is enabled that requires sophisticated row mapping (and in fact would break up if someone tries to navigate between dataSets using the Statement::previousDataSet etc.)
